src/data.rs

```diff
@@ -789,6 +809,254 @@ impl Drop for DataTree<'_> {
     }
 }
 
+// ===== impl DataTreeOwningRef =====
+
+impl<'a> DataTreeOwningRef<'a> {
+    unsafe fn from_raw(tree: DataTree<'a>, raw: *mut ffi::lyd_node) -> Self {
+        DataTreeOwningRef { tree, raw }
+    }
+
+    /// Create a new node or modify existing one in the data tree based on a
+    /// path.
+    ///
+    /// If path points to a list key and the list instance does not exist,
+    /// the key value from the predicate is used and value is ignored. Also,
+    /// if a leaf-list is being created and both a predicate is defined in
+    /// path and value is set, the predicate is preferred.
+    ///
+    /// For key-less lists and state leaf-lists, positional predicates can be
+    /// used. If no preciate is used for these nodes, they are always created.
+    ///
+    /// The output parameter can be used to change the behavior to ignore
+    /// RPC/action input schema nodes and use only output ones.
+    ///
+    /// Returns the last created or modified node (if any).
+    pub fn new_path(
+        context: &'a Context,
+        path: &str,
+        value: Option<&str>,
+        output: bool,
+    ) -> Result<Self> {
+        let mut tree = DataTree::new(context);
+        let raw = {
+            match tree.new_path(path, value, output)? {
+                Some(node) => node.raw,
+                None => tree.find_path(path)?.raw,
+            }
+        };
+        Ok(unsafe { DataTreeOwningRef::from_raw(tree, raw) })
+    }
+
+    pub fn noderef(&'a self) -> DataNodeRef<'a, 'a> {
+        DataNodeRef {
+            tree: &self.tree,
+            raw: self.raw,
+        }
+    }
+
+    fn _parse_op(
+        context: &'a Context,
+        raw: *mut ffi::lyd_node,
+        data: impl AsRef<[u8]>,
+        format: DataFormat,
+        op_type: ffi::lyd_type::Type,
+        op_node_ptr: *mut *mut ffi::lyd_node,
+    ) -> Result<()> {
+        let mut opaque = std::ptr::null_mut();
+        let opaque_ptr = &mut opaque;
+
+        // Create input handler.
+        let mut ly_in = std::ptr::null_mut();
+        let ret = if data.as_ref().is_empty() {
+            let blank = CString::new("").unwrap();
```

Comment:
Is this really necessary?

I just checked the source of `ly_in_new_memory()` and it does this:
```c
LIBYANG_API_DEF LY_ERR 
ly_in_new_memory(const char *str, struct ly_in **in)
{   
    LY_CHECK_ARG_RET(NULL, str, in, LY_EINVAL);
```

I believe you want this function to do nothing when `data` is empty, instead of returning an error. Could you confirm?

Reply:
Correct; however, this code is buggy as it doesn't allocate new memory it just keeps a pointer to the passed in data. So I need to fix it.